### ISSUE TITLE
mnist training E2E test failing; chief replica fails

### ISSUE THREAD

**Issue by jlewi** [kubeflow/examples#515]
The training step for mnist is failing.
Here's a failed run from a recent periodic run
http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-periodic-examples-mnist-90-70fd?tab=workflow&nodeId=kubeflow-periodic-examples-mnist-90-70fd-1605631456&sidePanel=logs%3Akubeflow-periodic-examples-mnist-90-70fd-1605631456%3Amain

The TFJob is in a failed status

```
   {
       "status": "True",
       "lastUpdateTime": "2019-02-22T10:27:54Z",
       "lastTransitionTime": "2019-02-22T10:27:54Z",
       "reason": "TFJobFailed",
       "message": "TFJob mnist-test-90 has failed because 1 Chief replica(s) failed.",
       "type": "Failed"
     }
   ],
```

Events indicate the chief pod was createad

```
'kind': None,
'last_timestamp': datetime.datetime(2019, 2, 22, 10, 27, 47, tzinfo=tzlocal()),
'message': 'Created pod: mnist-test-90-chief-0',
'metadata': {'annotations': None,
             'cluster_name': None,
             'creation_timestamp': datetime.datetime(2019, 2, 22, 10, 27, 48, tzinfo=tzlocal()),
             'deletion_grace_period_seconds': None,
             'deletion_timestamp': None,
             'finalizers': None,
             'generate_name': None,
             'generation': None,
             'initializers': None,
             'labels': None,
             'name': 'mnist-test-90.1585a8d24fb78cab',
             'namespace': 'mnist-90',
             'owner_references': None,
             'resource_version': '1397',
             'self_link': '/api/v1/namespaces/mnist-90/events/mnist-test-90.1585a8d24fb78cab',
             'uid': '8048d797-368c-11e9-932e-42010af0008a'},
'reason': 'SuccessfulCreatePod',
'related': None,
'reporting_component': '',
'reporting_instance': '',
'series': None,
'source': {'component': 'tf-operator', 'host': None},
'type': 'Normal'}
```
I couldn't find any logs for that pod though.

Event logs indicate pod exited with code 1 `'Pod: mnist-90.mnist-test-90-chief-0 exited with code 1',

**Comment by jlewi** [comment]
The missing logs might be due to kubeflow/kubeflow#2300

To debug it my suggestion would be to try to reproduce the test by running just the training test manually against a deployed instance of Kubeflow and see if that reproduces the error.

Another hack would at least for debugging be to tadd a 2 minute sleep at the start of the code just to see if that gives us logs.

**Comment by jinchihe** [comment]
Found same error in pod log, may be root course, but need to be confirm.

```
 "errors": [

 {

 "domain": "global",

 "reason": "forbidden",

 "message": "kf-v0-4-n00-user@kubeflow-ci.iam.gserviceaccount.com does not have storage.objects.get access to kubeflow-
ci_temp/mnist/models/1384."

 }

 ],
```

**Comment by jinchihe** [comment]
Seems the problem cannot be 100% reproduced. sometimes the testing can be passed. 

```
INFO|2019-02-24T07:03:43|/mnt/test-data-volume/kubeflow-examples-presubmit-mnist-516-ac54d27-1387-2986/src/kubeflow/tf-operator/py/tf_job
_client.py|115| Job mnist-test-1387 in namespace mnist-1387; uid=1c0e547a-3802-11e9-844a-42010af001b3; conditions=[u'Created', u'Running'
, u'Succeeded']
INFO|2019-02-24T07:03:43|tfjob_test.py|79| Final TFJob:
{
 "status": {
   "completionTime": "2019-02-24T07:03:32Z",
   "conditions": [
     {
       "status": "True",
       "lastUpdateTime": "2019-02-24T07:02:12Z",
       "lastTransitionTime": "2019-02-24T07:02:12Z",
       "reason": "TFJobCreated",
       "message": "TFJob mnist-test-1387 is created.",
       "type": "Created"
     },
     {
       "status": "False",
       "lastUpdateTime": "2019-02-24T07:02:51Z",
       "lastTransitionTime": "2019-02-24T07:02:51Z",
       "reason": "TFJobRunning",
       "message": "TFJob mnist-test-1387 is running.",
       "type": "Running"
     },
     {
       "status": "True",
       "lastUpdateTime": "2019-02-24T07:03:32Z",
       "lastTransitionTime": "2019-02-24T07:03:32Z",
       "reason": "TFJobSucceeded",
       "message": "TFJob mnist-test-1387 successfully completed.",
       "type": "Succeeded"
     }
```

**Comment by jlewi** [comment]
Thanks @jinchihe. This is really great sleuthing!

This seems like #953. But create_kf_instance.py should be calling delete_deployment.sh which should delete all role bindings
https://github.com/kubeflow/kubeflow/blob/610705224bc54a4ee79c981f349829dbc7277094/scripts/gke/delete_deployment.sh#L86

We should check the logs of create_kf_instance.py to see if there are problems occurring.

**Comment by jlewi** [comment]
I think a hack would be to decrease the frequency with which we redeploy so that the cluster will stay up for 1-2 days. This will make it easier to manually debug and fix. This should hopefully make it easier to get a green run in the presubmit.

/cc @gabrielwen

**Comment by jlewi** [comment]
I confirmed the service account has the storage admin role.
[iam.policy.yaml.txt](https://github.com/kubeflow/examples/files/2902965/iam.policy.yaml.txt)

**Comment by jinchihe** [comment]
@jlewi  @gabrielwen Thank you!  Seems the problem has been resolved, and `kubeflow-examples-presubmit` can be passed successfully now since the tfjobs finished normally.

@jlewi In the PR #516, printed out error logs from related trainning pods for debugging if the tfjobs failed, do you think is that needed? if so please help to review the #516 changes, otherwise I will close the PR. Thanks a lot!

**Comment by jinchihe** [comment]
@gabrielwen The problem happened again while running E2E for PR #531, Could you please take some actions as previous way? Thanks a lot!

```
 "errors": [
  {
   "domain": "global",
   "reason": "forbidden",
   "message": "kf-v0-4-n00-user@kubeflow-ci.iam.gserviceaccount.com does not have storage.objects.get access to kubeflow-ci_temp/mnist/mo
dels/1428."
  }
 ],
 "code": 403,
 "message": "kf-v0-4-n00-user@kubeflow-ci.iam.gserviceaccount.com does not have storage.objects.get access to kubeflow-ci_temp/mnist/mode
ls/1428."
}
}
```

**Comment by jinchihe** [comment]
@kunmingg Could you please help to take a look the problem? 
It is worth mentioning that maybe some things are done by @jlewi or someone before, so the E2E tests are successfully after that. The problem happened again recently.  Thanks!

```
kf-v0-4-n00-user@kubeflow-ci.iam.gserviceaccount.com does not have storage.objects.get access to kubeflow-ci_temp/mnist/models/xxx
```

**Comment by jlewi** [comment]
I suspect this is a variant of kubeflow/kubeflow#953 where the role binding wasn't deleted.

We might need to fix the auto-deploy script to properly delete the bindings between deployments.

@jinchihe The auto-deploy scripts are here:
https://github.com/kubeflow/testing/tree/master/test-infra/auto-deploy

It might be helpful to write a simple test. This could be something as simple as a shell script that tries to run `gsutil ls ...`. That might be helpful to reproduce and fix the problem.

**Comment by jinchihe** [comment]
@jlewi  I checked @kunmingg by Slack to resove the role binding issue since same limitations for trying gcp from my location :-). kunmingg is going to double check now. Thanks.

**Comment by jlewi** [comment]
We have an updated cluster that is being deployed in project kubeflow-ci-deployment
https://kf-vmaster-n00.endpoints.kubeflow-ci-deployment.cloud.goog/
cluster: kf-vmaster-n00
zone: us-east1-b

This cluster is being deployed with the new kfctl binary. I'm hoping this new binary takes care of the GCS issues.

Do you want to update the test to use the new cluster?

**Comment by jinchihe** [comment]
we can close the ticket now, the testing works fine now with below configuration. The configuration has been updated in PR #531. 

```
project: kubeflow-ci-deployment
zone: us-east1-b
cluster: kf-vmaster-n00
```

**Comment by gabrielwen** [comment]
could you wait for kubeflow/testing#354 and use functions there to find testing cluster to connect to?

**Comment by jlewi** [comment]
@jinchihe Has this issue been fixed? It looks like kubeflow/testing#354 has been merged. Should we open up a new issue to use it in the mnist E2E test to get the cluster to test on.

**Comment by jinchihe** [comment]
> @jinchihe Has this issue been fixed? It looks like [kubeflow/testing#354](https://github.com/kubeflow/testing/pull/354) has been merged. Should we open up a new issue to use it in the mnist E2E test to get the cluster to test on.

@jlewi Yes, I think we can close the issue now. Thanks.
